### PR TITLE
Propagate response headers in case of invalid response

### LINKED ISSUE #24703: Response Headers not propagated to the filter state when response is invalid
According to the description of the propagate_response_headers config:
https://github.com/envoyproxy/envoy/blob/23eacb181632df5e7be61f00a09e78bdfc686482/api/envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto#L104
We expect to have response headers saved to the downstream info filter state for consumption by the network filters in all cases, but I see in the implementation that in case of invalid response or end of stream the headers are not propagated:
https://github.com/envoyproxy/envoy/blob/23eacb181632df5e7be61f00a09e78bdfc686482/source/common/tcp_proxy/upstream.h#L173
Is there any reason not to propagate headers in all cases?
@kyessenov ?

### PR DESCRIPTION
**Pull request by 0xquark**
Signed-off-by: Karanjot Singh <thestrangequarks@protonmail.com>

Commit Message: Propagate response headers in case of invalid response
Additional Description: In case of invalid response or end of stream the headers are not propagated ( Fixes #24703 )
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/common/tcp_proxy/upstream.h
Comment (IssaAbuKalbein) [inline source/common/tcp_proxy/upstream.h:176]: I suggest the change to be the following:
      bool isValidResponse = parent_.isValidResponse(*headers);
      parent_.config_.propagateResponseHeaders(std::move(headers),
                                               parent_.downstream_info_.filterState());
      if (!isValidResponse || end_stream) {...}
Because we use std::move(headers) in propagateResponseHeaders

Comment (0xquark) [reply source/common/tcp_proxy/upstream.h]: Thank you! Done

**Review by ohadvano — DISMISSED** [review]

**Comment by 0xquark** [comment]
Hi @yanavlasov I am new to writing unit and integration tests. Could you help me a bit getting started?

**Comment by kyessenov** [comment]
@0xquark The test that needs to be modified is `test/integration/tcp_tunneling_integration_test.cc`.

**Comment by alyssawilk** [comment]
/wait on CI.
cc @kyessenov I don't think this needs runtime guarding but I'd like your thoughts

File: test/common/tcp_proxy/upstream_test.cc
Comment (kyessenov) [inline test/common/tcp_proxy/upstream_test.cc:187]: here and elsewhere, please remove `this->`, it's not needed

**Review by kyessenov — CHANGES_REQUESTED** [review]
Yeah, I think it's fine without a guard. Please fix the test.

**Comment by ohadvano** [comment]
Can we add this fix to the upcoming Envoy release?

**Comment by 0xquark** [comment]
I've done the requested changes! 
@kyessenov @ohadvano can you review now ?
Apologies for the delay, was out last week!

**Review by ohadvano — DISMISSED** [review]

**Comment by ohadvano** [comment]
Thank you @0xquark, looks good to me

**Comment by IssaAbuKalbein** [comment]
@0xquark I tried to push this integration test to your PR, can you add this test, please?
https://github.com/0xquark/envoy/pull/1

File: test/common/tcp_proxy/upstream_test.cc
Comment (IssaAbuKalbein) [inline test/common/tcp_proxy/upstream_test.cc:191]:
```suggestion
  EXPECT_CALL(*conn_pool_callbacks_raw, onFailure()).Times(0);
```

File: test/common/tcp_proxy/upstream_test.cc
Comment (IssaAbuKalbein) [inline test/common/tcp_proxy/upstream_test.cc:192]:
```suggestion
  EXPECT_CALL(*conn_pool_callbacks_raw, onSuccess(_)).Times(0);
```

File: test/common/tcp_proxy/upstream_test.cc
Comment (IssaAbuKalbein) [inline test/common/tcp_proxy/upstream_test.cc:196]:
```suggestion

```

**Review by IssaAbuKalbein — CHANGES_REQUESTED** [review]
Can you, please, fix format issues

**Comment by 0xquark** [comment]
Done @IssaAbuKalbein

File: test/common/tcp_proxy/upstream_test.cc
Comment (IssaAbuKalbein) [inline test/common/tcp_proxy/upstream_test.cc:195]:
```suggestion
}

```

**Review by IssaAbuKalbein — CHANGES_REQUESTED** [review]

File: test/common/tcp_proxy/upstream_test.cc
Comment (IssaAbuKalbein) [inline test/common/tcp_proxy/upstream_test.cc:196]:
```suggestion

```

**Review by IssaAbuKalbein — CHANGES_REQUESTED** [review]
Please, remove the space in the line (do not remove the line itself)

File: test/common/tcp_proxy/upstream_test.cc
Comment (0xquark) [reply test/common/tcp_proxy/upstream_test.cc:196]: I apologize if there is any confusion on my part. I noticed in a previous 2 commit, the line of code was removed and then added back again since you requested me to add it.

Comment (IssaAbuKalbein) [reply test/common/tcp_proxy/upstream_test.cc:196]: No problem, I meant to remove the redundant space in the line not the line itself

Comment (0xquark) [reply test/common/tcp_proxy/upstream_test.cc:196]: Upon reviewing the code on my end, I am unable to identify any redundant spaces. However, when I attempt to commit the changes, it appears that there is an issue with unnecessary spaces.

Comment (0xquark) [reply test/common/tcp_proxy/upstream_test.cc]: I think it should work now!

File: test/common/tcp_proxy/upstream_test.cc
Comment (IssaAbuKalbein) [inline test/common/tcp_proxy/upstream_test.cc:196]: Can we remove this test? I think the integration test is good enough.
Also , I see that it's exactly the same as  previous test "OnFailureCalledOnInvalidResponse".

Comment (IssaAbuKalbein) [reply test/common/tcp_proxy/upstream_test.cc:196]: @0xquark ?

Comment (0xquark) [reply test/common/tcp_proxy/upstream_test.cc]: Done!

**Comment by IssaAbuKalbein** [comment]
@alyssawilk, @kyessenov Can you take a look now after the changes?

**Review by kyessenov — DISMISSED** [review]
Thanks for fixing the test.

**Comment by IssaAbuKalbein** [comment]
@0xquark Can you, please, merge the PR?

**Comment by 0xquark** [comment]
@IssaAbuKalbein I am not the maintainer of this repo.

**Review by ohadvano — DISMISSED** [review]

**Review by IssaAbuKalbein — APPROVED** [review]

**Comment by IssaAbuKalbein** [comment]
@kyessenov @alyssawilk Can you merge the PR, please?

File: source/common/tcp_proxy/upstream.h
Comment (alyssawilk) [inline source/common/tcp_proxy/upstream.h:174]: can you please rename to is_valid_response?

Comment (IssaAbuKalbein) [reply source/common/tcp_proxy/upstream.h:174]: @0xquark

Comment (IssaAbuKalbein) [inline source/common/tcp_proxy/upstream.h:174]:
```suggestion
      bool is_valid_response = parent_.isValidResponse(*headers);
```

File: source/common/tcp_proxy/upstream.h
Comment (IssaAbuKalbein) [inline source/common/tcp_proxy/upstream.h:177]:
```suggestion
      if (!is_valid_response || end_stream) {
```

**Review by IssaAbuKalbein — CHANGES_REQUESTED** [review]

File: source/common/tcp_proxy/upstream.h
Comment (0xquark) [reply source/common/tcp_proxy/upstream.h]: Done

**Review by kyessenov — APPROVED** [review]

**Pull request merged**
